pkg/k8s/controllers/reconciler.go

```diff
@@ -0,0 +1,276 @@
+		log.WithError(getErr).Errorf("unable to get %s CR", r.Kind())
+		// this is likely a controller permission issue
+		rr.ControllerRes.RequeueAfter = MaxRequeueInterval()
+		rr.ReconcileErr = getErr
+		return &rr
+	}
+
+	// Status changes trigger Reconcile events, ignore those if spec did not change.
+	var hash string
+	hash, rr.ReconcileErr = in.GetSpec().Hash()
+	if rr.ReconcileErr != nil {
+		log.WithError(rr.ReconcileErr).Error("failed to calculate hash")
+		// this is very likely a permanent error, no need to retry too frequently
+		// still, if new deployment happens, we want the new deployment to retry this CR
+		rr.ControllerRes.RequeueAfter = MaxRequeueInterval()
+		return &rr
+	}
+
+	rr.Skipped = !in.GetStatus().ShouldReconcile(hash, log)
+	if rr.Skipped {
+		// accurate skip reason is logged by ShouldReconcile
+		return &rr
+	}
+
+	rr = r.handler.Reconcile(ctx, log, in)
+	if rr.Skipped {
+		// do not take any action here if impl asked to skip status+hash updates (maybe CR is meant for a diff bento)
+		// endReconcile is still called - logging done inside
+		return &rr
+	}
+
+	updateErr := r.updateStatus(ctx, log, in, &rr)
+	if updateErr != nil {
+		// If we fail to update status of the CR, this is very likely a permission issue.
+		// Override handler's decision in this case and proceed with slow retry.
+		rr.ReconcileErr = updateErr
+		rr.ControllerRes.RequeueAfter = MaxRequeueInterval()
+	}
+
+	// logging done inside endReconcile
+	return &rr
+}
+
+// updateStatus is called to update CR status based on reconcileErr
+func (r *Reconciler) updateStatus(
+	ctx context.Context,
+	log logrus.FieldLogger,
+	in resources.Resource,
+	rr *ReconcileResult,
+) error {
+	// The spec is very unlikely to be changed by the controller, yet recalculate the hash just in case.
+	hash, err := in.GetSpec().Hash()
+	if err != nil {
+		log.WithError(err).Errorf("failed to calculate hash for the %s", r.Kind())
+		// logging done inside endReconcile
+		return err
+	}
+
+	// update the CR's status with the reconcile hash and reconcileErr (if present)
+	in.GetStatus().Update(hash, rr.ReconcileErr)
+	// capture reconcile fail count so far on this hash
+	rr.failCount = in.GetStatus().ReconcileFailCount
+
+	err = r.Client().Status().Update(ctx, in)
+	if err != nil {
+		log.WithError(err).Errorf("unable to update status for %s CR: %+v", r.Kind(), in.GetStatus())
+	}
+
+	return err
+}
+
+// Setup registers this Reconciler instance as a controller to process target resources.
+func (r *Reconciler) Setup(mgr ctrl.Manager) error {
+	err := ctrl.NewControllerManagedBy(mgr).
+		Named(r.Kind()).
+		For(r.handler.CreateResource()).
+		Complete(r)
+	if err != nil {
+		r.log.WithError(err).Errorf("failed to setup Reconciler as a controller for %s resource with k8s manager", r.Kind())
+	}
+	return err
+}
+
+// getRequeueDuration returns the requeue interval to retry, based on number of times this CR failed so far.
+func getRequeueDuration(failCount int) time.Duration {
```

Comment:
golang naming convention: s/get//g

```suggestion
func requeueDuration(failCount int) time.Duration {
```

Reply:
ack, will push shortly

Reply:
done